File: src/main/python/NerDataset.py
import torch
from torch.utils.data import Dataset

from NerVocab import mask

class NerDataset(Dataset):
    def __init__(self, dataset_path, vocab):
        self.samples = []
        self.labels = {}
        self.vocab = vocab

        # Load data and construct label map
        with open(dataset_path, "r") as data_file:
            cur_sentence = [[], []]
            for line in data_file:
                cur = line.strip()
                if cur == "":
                    self.samples.append(cur_sentence)
                    cur_sentence = [[], []]
                else:
                    token, label = cur.split()
                    cur_sentence[0].append(token)
                    cur_sentence[1].append(label)

                    if label not in self.labels:
                        # We will sort and replace number later
                        self.labels[label] = -1

        for i, label in enumerate(sorted(self.labels.keys())):
            self.labels[label] = i

    def __len__(self):
        return len(self.samples)

    def __getitem__(self, index):
        assert len(self.samples) > index and index >= 0
        sentence, labels = self.samples[index]

        out_sent = []
        for token in sentence:
            if token in self.vocab:
                out_sent.append(self.vocab[token])
            else:
                out_sent.append(self.vocab[mask])

        out_labels = [self.labels[label] for label in labels]

        out_sent = torch.tensor(out_sent)
        out_labels = torch.tensor(out_labels)

        return (out_sent, out_labels)

File: src/main/scala/com/keithalcock/torchscript/NerDataset.scala
package com.keithalcock.torchscript

import org.pytorch.Tensor

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.io.Source

class NerDataset(datasetPath: String, val vocab: Map[String, Int]) extends IndexedSeq[(Tensor, Tensor)] {

  case class Sample(tokens: Array[String], labels: Array[String])

  val samples = new ArrayBuffer[Sample]()
  val labels = mutable.Map.empty[String, Int]

  {
    // Return the samples and labels from this?
    val tokens = new ArrayBuffer[String]()
    val labels = new ArrayBuffer[String]()
    val source = Source.fromFile(datasetPath)

    source.getLines.foreach { line =>
      val cur = line.trim

      if (cur.isEmpty) {
        samples.append(Sample(tokens.toArray, labels.toArray))
        tokens.clear()
        labels.clear()
      }
      else {
        val Array(token, label) = cur.split(' ')
        tokens.append(token)
        labels.append(label)

        // Make this one a set and just add
        if (!this.labels.contains(label))
          this.labels(label) = -1
      }
    }
    source.close

    this.labels.keys.toSeq.sorted.zipWithIndex.foreach { case (key, index) =>
      this.labels(key) = index
    }
  }

  override def length: Int = samples.length

  override def apply(index: Int): (Tensor, Tensor) = {
    require(0 <= index && index < length)
    val sample = samples(index)
    val tokenIndexes = sample.tokens.map { token =>
      vocab.getOrElse(token, vocab(NerVocab.mask))
    }
    val labelIndexes = sample.labels.map { label =>
      labels(label)
    }

    (
      Tensor.fromBlob(tokenIndexes, Array(1L, tokenIndexes.length.toLong)),
      Tensor.fromBlob(labelIndexes, Array(1L, labelIndexes.length.toLong))
    )
  }
}
